---
title: '`flippant` -- An `R` package for the automated analysis of dithionite scramblase
  assays'
author: Johannes Graumann ^†,^*, Birgit Ploier ^‡^, Richard J. Cotton ^†^, Mike A.
  Goren ^‡^ and Anant K. Menon ^‡^
date: "September 19, 2015"
output:
#  word_document
  pdf_document:
    fig_caption: yes
csl: bioinformatics.csl
bibliography: flippant.bib
---

†
:    \
    Weill Cornell Medical College in Qatar, P.O.Box 24144, Doha, State of Qatar

‡
:   Department of Biochemistry, Weill Cornell Medical College, New York, NY 
    10065, USA

\*
:   To whom correspondence should be addressed: jog2030@qatar-med.cornell.edu

```{r, echo = FALSE}
library(flippant, quietly = TRUE)
library(grid, quietly = TRUE)
library(magrittr, quietly = TRUE)
library(multipanelfigure, quietly = TRUE)
library(stringi, quietly = TRUE)
flippant_dependencies <- packageDescription("flippant",fields=c("Depends","Imports")) %>%
  unlist(use.names = FALSE) %>%
  stri_split(regex = ",", omit_empty = TRUE) %>%
  unlist(use.names = FALSE) %>%
  stri_replace_all(replacement = "", regex = "\\s*\\(.*\\)\\s*", omit_empty = TRUE) %>%
  stri_replace_all(replacement = "", regex = "\\s*", omit_empty = TRUE) %>%
  tail(n=length(.) - 1) %>%
  sort() %>%
  ( function(x){
    paste0(
      c(
        "`",
        paste0(head(x, n = length(x) - 1), collapse = "`, `"),
        "` and `",
        tail(x, n = 1),
        "`"),
      collapse = "")} )

tmpDir <- tempdir()
file.path("extdata", "PloierEtAl_Data.zip") %>%
  system.file(
    package = "flippant",
    mustWork = TRUE) %>%
  unzip(
    overwrite = TRUE,
    exdir = tmpDir)

figure1 <- multipanelfigure(
  width = 17.22, 
  height = 18, 
  units = "cm", 
  columns = 2, 
  rows = 2)

knitr::opts_chunk$set(
  fig.width = 6.78, 
  fig.height = 7.09, 
  warning = FALSE, 
  message = FALSE)
```

#Abstract

##Summary

##Availability and Implementation

##Contact
jog2030@qatar-med.cornell.edu

##Supplementary Information

# Introduction
Scramblases are proteins mediating the mass-action driven equilibration of 
lipids between the leaflets of lipid bilayers that constitute biological membranes
[@waechter_flippases_2013]. Independent of energy equivalents such as ATP they 
facilitate this ``flipping'' by providing a mechanism of transition through the 
hydrophobic environment of membranes to polar lipid head groups, thus reducing 
the massive energetic cost that process would carry otherwise 
[@kornberg_inside-outside_1971]. The underlying molecular mechanisms remain 
largely enigmatic [for a discussion including exceptions see @ernst_phospholipid_2015].
Proteins supporting flipping processes as central as the equilibration of the 
ubiquitous phosphoglycerolipids across the membrane of the endoplasmic 
reticulum, where they are synthesized in the cytoplasmic monolayer, have yet to 
be identified **(I was looking for a recent 
citation that clearly spelled that out - in vain. Do you have a proposal?)**. 
None the less, there has been much recent progress in scramblase identification 
and characterization. Proteins shown to have scramblase activity include 
FtsW [@mohammadi_identification_2011], opsin/rhodopsin
[@menon_opsin_2011;@goren_constitutive_2014], TMEM16 
[@malvezzi_ca2-dependent_2013], $\beta$ 2-adrenergic and adenosine A2A 
receptors [both @goren_constitutive_2014]. Extrapolating from the latter findings 
and the opsin/rhodopsin case, we speculate that lipid scrambling may, in fact, 
be an intrinsic property of G-protein coupled receptors in general 
[@goren_constitutive_2014].

To interrogate scramblases as a functional class of proteins, we have developed 
a biochemical assay suited to characterize scramblase activity of purified 
proteins and protein extracts based on the reconstitution of scramblases or 
candidate proteins into preformed synthetic liposomes that contain trace amounts
of nitrobenzoxadiazole (NBD) labeled fluorescent lipids [@chang_chemical_2004].
Upon addition of membrane-impermeable reducing agents such as dithionite to 
these liposomes, fluorophores in the external leaflet are quenched and fluorescence in scramblase-free liposomes accordingly quickly drops to approximately 50% of the 
initial value. However, for proteoliposomes reconstituted with proteins 
conferring a scramblase activity, fluorescence disappears entirely, as
fast scramblase-mediated equilibration between outer (exposed) and inner
(protected) leaflet of the bilayer renders all fluorescent lipids
accessible to reduction. By titration of the protein amount reconstituted into
the liposome membranes it is thus possible to characterize scramblases and probe
their activity. Beyond the identification of scramblase activities in crude 
membrane preparations [@chang_chemical_2004;@rajasekharan_flip-flop_2011] and 
functional validation of candidate scramblases [@menon_opsin_2011;@malvezzi_ca2-dependent_2013;@goren_constitutive_2014], the 
assay has also been used to characterize the stoichiometry of scramblase 
homo-multimers reconstituting into the proteoliposomes 
[@ploier_perturbed_submitted] and probe the ATP-dependence of flippases, 
molecules that under energy-consumption flip lipids against a concentration 
gradient between bilayer leaflets [@soupene_atp8a1_2008].
**Are you aware of further use? More adoption would fortify the need for the 
package ...**

Starting with the identification of baseline and post-reduction fluorescence 
from fluorometer-generated spectra, via the calculation of the underlying 
statistics and through publication-grade representation of the results, data 
analysis in the context of this assay has until now involved an extensive series
of manual steps, at most supported by spreadsheet calculation facilities
[see @menon_opsin_2011;@goren_constitutive_2014 for extensive step-wise discussion].
Based on the `R` free software environment for statistical computing and 
graphics [@r_core_team_r:_2015] and prominently drawing on the power and 
elegance of the `ggplot2` plotting system [@wickham_ggplot2_2009], we here 
provide `flippant`, a package that comprehensively, fast and reproducibly covers
all data analysis and graphing needs from raw fluorescence spectra produced by 
scramblase assays. A simple parametrizing spreadsheet is the sole required user 
input.

#Installation
`flippant` has been integrated into the `bioconductor` collection of tools for 
biomedical data analysis [@huber_orchestrating_2015] and is thus install- and
loadable as follows from within a functional `R` environment:
```{r, eval = FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite(pkgs = "flippant")
library(flippant)
```
The package depends on the following packaged extensions to `R` base 
functionality, which are automatically satisfied on installation: 
`r flippant_dependencies`.

#Input Data and Data Processing
To analyse raw data stemming from dithionite scramblase assays, `flippant` 
requires for each data point information on the path to the 
fluorometer-generated spectrum, the amount of protein used to reconstitute 
proteoliposomes, the assay volume before and after addition of 
dithionite (to calculate a volume correction factor), the amount of lipids 
present (to calculate a protein to phospholipid ratio or PPR) etc. The user 
supplies this information in a spreadsheet (see below for supporting 
functionality).

Spectral input data is read by `flippant` using the file paths provided in the
spreadsheet. Raw spectra produced by QuantaMaster fluorometer instruments 
(Photon Technology International, Inc., Edison, New Jersey) running software 
versions FelixGX v4.1 and Felix32 v1.20, respectively, as well as a simple 
generic tab-delimited spectrum representing file format with time (in seconds) 
and fluorescence intensity representing columns are currently supported. 
File type determination is handled algorythmically.

Based on @menon_opsin_2011 and @goren_constitutive_2014, data is processed as 
follows. Input is format checked and defaults are injected for facultative 
parameters/columns as appropriate. Fluorescence spectra are parsed. This includes 
algorithmic determination of when dithionite was added to the sample using peak
detection by continuous wavelet transformation as implemented in the 
`wmtsa` `R` package [@percival_wavelet_2000;@constantine_wmtsa:_2013]. 
Acquisition time is aligned for all spectra in the data set such that the zero 
time point henceforth corresponds to the time of addition. 
Pre-dithionite-addition baseline fluorescence $F_{\text{baseline}}$ is 
determined for each spectrum by averaging (median) over the ten values preceding
dithionite addition. Post-dithionite-addition minimum fluorescence 
$F_{\text{min}}$ is analogously calculated from the last ten data points before 
(and including) a default 400 s or the time point of measurement supplied by the
user in the parameter spreadsheet. $F_{\text{min}}$ is volume-corrected based on
supplied reaction volumes with and without dithionite and for each spectrum/data
point a fluorescence reduction is calculated as follows:
$$
y_{\text{raw}} = 1 - \frac{F_{\text{min}}}{F_{\text{baseline}}}
$$
A relative fluorescence reduction $y$ is calculated scaled to a (required) liposomes-only/no-protein control and the probability $P_{\geq 1}$ for a 
liposome to contain at least one scramblase molecules is calculated using
$$
P_{\geq 1} = \frac{(y - y_0)}{(y_{\text{max}} - y_0)}
$$
where $y$ is the relative fluorescence reduction for the proteoliposome 
measurement and $y_0$ for a measurement of protein-free liposomes. 
$y_{\text{max}}$ refers to either the maximal $y$ in the experiment or is 
derived from a monoexponential fit to the data (default) as a precaution for the 
case where the protein/phospholipid titration did not reach the plateau of the 
saturation curve. The behavior is adjustable using the `scaleTo` parameter of 
the corresponding functions (see below). Next a monoexponential curve is fitted
to the relationship of $P_{\geq 1}$ to the protein to phospholipid ratio (PPR; 
(mg/mmol)), providing the scramblase-characterizing fit factor $\alpha$.

#Main Features
##Template Generation for Data Input
Supporting the user in providing the input data needed, the following function 
generates a spreadsheet program-compatible tab-delimited `ASCII` template, 
including column names, commentary, expected data type and the default used when
omitted.
```{r, code=formatR::usage('scramblaseAssayInputTemplate'), eval=FALSE}
```
The application agnostic tab-delimited format must be maintained for the 
completed table, which is subjected to extensive input checks when read by 
`flippant`.

##Graphic Representation of Spectral Traces
In a first, graphical data analytic step of a scramblase assay-derived data 
set, fluorescent spectra or traces may be plotted in publication-appropriate
quality with acquisition time in the x- and fluorescence intensity on the y-axis
using
```{r, code = formatR::usage('scramblaseAssayTraces'), eval = FALSE}
```
and handing the function the path to the user assembled spreadsheet of input 
data (`x`). The `adjust` argument to the function call serves to selectively 
inactivate `wmtsa`-provided (see above) alignment of multiple spectra to the 
time of dithionite addition. Acquisition time coverage of the plot may be tuned 
by user-provided minimal (`timeMin` ) and maximal (`timeMax`) time points to be
included.

##Scramblase Assay Analysis
Plotting protein to phospholipid ratio (PPR) against the likelihood for a given
proteoliposome to contain one or more scramblases, the PPR plot is at the heart 
of the analysis of dithionite scramblase assays, allowing for the visual 
comparison of specific scramblase activity between interrogated experimental 
series, which may e.g. be representative of protein preparations including 
mutant protein forms or crude extracts with or without depletion of a candidate
scramblase. 
```{r, code = formatR::usage('scramblaseAssayPlot'), eval = FALSE}
```
allows for the complete generation of the PPR plot in publication quality from 
the user assembled spreadsheet of input data (`x`), including automatically 
added graphical representations of $\tau$, the PPR value equal to the fit 
constant $\alpha$, which implies all liposomes on average carrying one 
scramblase and 63% have one or more (i.e. are active). The behavior of the 
function may be adjusted to use scramblase-mediated fluorescence reduction 
normalized to either the maximum measurement in the series (`scaleTo = "data"`)
or the plateau of an exponential fit to it (`scaleTo = "model"`; for reasoning
see above) and whether or not to require the exponential curve fitted to the 
data to go through the origin.

Numeric scramblase assay results such as $\tau$ may be generated in tabular form
using
```{r, code = formatR::usage('scramblaseAssayStats'), eval = FALSE}
```
with the same modifying arguments to the function as for `scramblaseAssayPlot`.

#Case Study
*Retinitis pigmentosa* is a degenerative disease of the retina, the majority of 
cases of which is linked to mutant forms of the G-protein coupled receptor 
rhodopsin [@mendes_mechanisms_2005]. Motivated by our recent discovery that 
opsin/rhodopsin has scramblase activity [@menon_opsin_2011;@goren_constitutive_2014],
we hypothesized that a class of enigmatic rhodopsin mutations, which are known
to cause *retinitis pigmentosa* yet do not display the commonly associated 
molecular phenotypes of impaired binding of retinal, folding and/or trafficking or 
transducin activation, may in fact be impacted in their scramblase activity.
This hypothesis was disproved when tested using the dithionite scramblase assay
among other techniques. However, the same set of experiments pointed at a defect 
in rhodopsin dimerization prior to insertion into liposomes, generating the new 
hypothesis that the mutations may cause disease by interfering with the highly 
ordered quarternary structure of the molecule in the retina 
[@ploier_perturbed_submitted].

`flippant`s capabilities are demonstrated here by reanalyzing the data underlying 
Figure 5b in @ploier_perturbed_submitted. The following work flow in a first 
step plots spectral traces for all points of the data set (Figure 1A). Note that
alorithmic alignment to the time of dithionite addition is employed (default) as
well as custom trimming of the time axis to account for divergent acquisition 
times between experiments.
<!-- Fix F220CEx1_E.txt to F220CEx1_D.txt -->
```{r, eval = FALSE}
# Plot spectral traces (Figure 1A)
scramblaseAssayTraces("inputTable.txt", timeMax = 350)
```
```{r, echo = FALSE, warning=FALSE, message=FALSE, fig.cap = "**`flippant`-based reanalysis of data in Figure 5b from @ploier_perturbed_submitted.** (A) represents spectral traces as produced by a call to the function `scramblaseAssayTraces`."}
setwd(tmpDir)
figure1 <- addpanel(
  figure = figure1,
  topPanel = 1, 
  leftPanel = 1, 
  rightPanel = 2, 
  panel = scramblaseAssayTraces("inputTable.txt", timeMax = 350)
)
grid.draw(figure1)
```


#Conclusion

#Funding
J.G., R.J.C. and the Proteomics Core at WCMC-Q are supported by ``Biomedical 
Research Program'' funds at Weill Cornell Medical College in Qatar, a program 
funded by Qatar Foundation. B.P. was supported by a grant from Qatar National 
Research Fund’s National Priority Research Program (5-669-1-112) to J.G. and
A.K.M., as well as by the Austrian Science Fund (FWF), project J3686-B21.

#Acknowledgements

#References
