---
title: '*flippant* -- An R package for the automated analysis of dithionite scramblase
  assays'
author: Johannes Graumann ^†,^*, Birgit Ploier ^‡^, Richard J. Cotton ^†^,
  Michael A. Goren ^‡^ and Anant K. Menon ^‡^
date: '`r format(Sys.Date(), "%B %d,%Y")`'
output:
  pdf_document:
    fig_caption: yes
  word_document:
    fig_caption: yes
  html_document:
    fig_caption: yes
csl: bioinformatics.csl
bibliography: flippant.bib
---

†
:    \
    Research Division, Weill Cornell Medical College in Qatar, P.O.Box 24144, Doha, State of Qatar

‡
:   Department of Biochemistry, Weill Cornell Medical College, New York, NY 
    10065, USA

\*
:   To whom correspondence should be addressed: jog2030@qatar-med.cornell.edu

Current Git Hash
:   `r system("git log -1 --format='%H'", intern = TRUE)`

Git Clean
:   `r ifelse(length(system("git status -s", intern = TRUE)) == 0, "TRUE", "FALSE")`

```{r, echo = FALSE}
library(flippant, quietly = TRUE)
library(grid, quietly = TRUE)
library(magrittr, quietly = TRUE)
library(microbenchmark, quietly = TRUE)
library(multipanelfigure, quietly = TRUE)
library(stringi, quietly = TRUE)
flippant_dependencies <- packageDescription("flippant",fields=c("Depends","Imports")) %>%
  unlist(use.names = FALSE) %>%
  stri_split(regex = ",", omit_empty = TRUE) %>%
  unlist(use.names = FALSE) %>%
  stri_replace_all(replacement = "", regex = "\\s*\\(.*\\)\\s*", omit_empty = TRUE) %>%
  stri_replace_all(replacement = "", regex = "\\s*", omit_empty = TRUE) %>%
  tail(n=length(.) - 1) %>%
  sort() %>%
  ( function(x){
    paste0(
      c(
        "*",
        paste0(head(x, n = length(x) - 1), collapse = "*, *"),
        "* and *",
        tail(x, n = 1),
        "*"),
      collapse = "")} )

tmpDir <- tempdir()
file.path("extdata", "PloierEtAl_Data.zip") %>%
  system.file(
    package = "flippant",
    mustWork = TRUE) %>%
  unzip(
    overwrite = TRUE,
    exdir = tmpDir)

figure1 <- multipanelfigure(
  widths = rep(17.22/3-2*0.25, 3), 
  heights = rep(27/7-5*0.25, 7), 
  units = "cm",
  interPanelSpacing = 0.25)

knitr::opts_chunk$set(
  fig.width = 6.78, 
  fig.height = 10.63, 
  warning = FALSE, 
  message = FALSE)
```

#Abstract
Dithionite scramblase assays are an established means of probing the lipid 
scrambling activity of protein extracts and purified proteins. While the assay
has yielded insight into the scramblase activity in crude membrane preparations,
functional validation of candidate scramblases, stoichiometry of scramblase 
complexes and probe ATP-dependence of flippases, data analysis in its context 
has remained a task involving many manual steps. With the extension package 
*flippant* to R, a free software environment for statistical computing and 
graphics, we introduce an integrated solution for the analysis and 
publication-grade graphical presentation of dithionite scramblase assays and 
demonstrate its utility in revisiting an originally manual analysis from the 
publication record, closely reproducing the reported results.

# Introduction
Scramblases are proteins mediating the bidirectional, mass-action driven 
equilibration of lipids between the leaflets of lipid bilayers that constitute 
biological membranes [@waechter_flippases_2013]. Independent of energy 
equivalents such as ATP they facilitate this ``flipping'' by providing a 
mechanism of transition through the hydrophobic environment of membranes to 
polar lipid head groups, thus reducing the massive energetic cost that process 
would carry otherwise [@kornberg_inside-outside_1971]. The underlying molecular 
mechanisms remain largely enigmatic [for a discussion including exceptions see @ernst_phospholipid_2015].
Proteins supporting flipping processes as central as the equilibration of the 
ubiquitous phosphoglycerolipids across the membrane of the endoplasmic 
reticulum, where they are synthesized in the cytoplasmic monolayer, have yet to 
be identified [@menon_specific_2000]. 
None the less, there has been much recent progress in scramblase identification 
and characterization. Proteins shown to have scramblase activity include 
FtsW [@mohammadi_identification_2011], opsin/rhodopsin
[@menon_opsin_2011;@goren_constitutive_2014], TMEM16 
[@malvezzi_ca2-dependent_2013], $\beta$ 2-adrenergic and adenosine A2A 
receptors [both @goren_constitutive_2014]. Extrapolating from the latter findings 
and the opsin/rhodopsin case, one may speculate that lipid scrambling may, in
fact, be an intrinsic property of G-protein coupled receptors in general 
[@goren_constitutive_2014].

To interrogate scramblases as a functional class of proteins, a biochemical
assay was developed suited to characterize scramblase activity of purified 
proteins and protein extracts based on the reconstitution of scramblases or 
candidate proteins into preformed synthetic liposomes that contain trace amounts
of nitrobenzoxadiazole (NBD) labeled fluorescent lipids [@chang_chemical_2004].
Upon addition of membrane-impermeable reducing agents such as dithionite to 
these liposomes, fluorophores in the external leaflet are quenched and fluorescence in scramblase-free liposomes accordingly quickly drops to approximately 50% of the 
initial value. However, for proteoliposomes reconstituted with proteins 
conferring a scramblase activity, fluorescence disappears entirely, as
fast scramblase-mediated equilibration between outer (exposed) and inner
(protected) leaflet of the bilayer renders all fluorescent lipids
accessible to reduction. By titration of the protein amount reconstituted into
the liposome membranes it is thus possible to characterize scramblases and probe
their activity. Beyond the identification of scramblase activities in crude 
membrane preparations [@chang_chemical_2004;@rajasekharan_flip-flop_2011] and 
functional validation of candidate scramblases [@menon_opsin_2011;@malvezzi_ca2-dependent_2013;@goren_constitutive_2014], the 
assay has also been used to characterize the stoichiometry of scramblase 
homo-multimers reconstituting into the proteoliposomes 
[@ploier_perturbed_submitted] and probe the ATP-dependence of flippases, 
molecules that under energy-consumption flip lipids against a concentration 
gradient between bilayer leaflets [@soupene_atp8a1_2008]. 

Starting with the identification of baseline and post-reduction fluorescence 
from fluorometer-generated spectra, via the calculation of the underlying 
statistics and through publication-grade representation of the results, data 
analysis in the context of this assay has until now involved an extensive series
of manual steps, at most supported by spreadsheet calculation facilities
[see @menon_opsin_2011;@goren_constitutive_2014 for extensive step-wise discussion].
Based on the R free software environment for statistical computing and 
graphics [@r_core_team_r:_2015] and prominently drawing on the power and 
elegance of the `ggplot2` plotting system [@wickham_ggplot2_2009], we here 
provide *flippant*, a package that comprehensively, fast and reproducibly covers
all data analysis and graphing needs from raw fluorescence spectra produced by 
scramblase assays. A simple parametrizing spreadsheet is the sole required user 
input.

#Availablility and Implementation
*flippant* has been implemented in R [@r_core_team_r:_2015] and integrated 
into the `bioconductor` collection of tools for biomedical data analysis [@huber_orchestrating_2015]. It is thus install- and loadable as follows from 
within a functional R environment:

```{r, eval = FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite(pkgs = "flippant")
library(flippant)
```

The package depends on the following packaged extensions to R base 
functionality, which are automatically satisfied on installation: 
`r flippant_dependencies`.

#Input Data and Data Processing
To analyse raw data stemming from dithionite scramblase assays, *flippant* 
requires for each data point information on the path to the 
fluorometer-generated spectrum, the amount of protein used to reconstitute 
proteoliposomes, the assay volume before and after addition of 
dithionite (to calculate a volume correction factor), the amount of lipids 
present (to calculate a protein to phospholipid ratio or PPR) etc. The user 
supplies this information in a spreadsheet (see below for supporting 
functionality).

Spectral input data is read by *flippant* using the file paths provided in the
spreadsheet. Raw spectra produced by QuantaMaster fluorometer instruments 
(Photon Technology International, Inc., Edison, New Jersey) running software 
versions FelixGX v4.1 and Felix32 v1.20, respectively, as well as a simple 
generic tab-delimited spectrum representing file format with time (in seconds) 
and fluorescence intensity representing columns are currently supported. The 
latter is trivially assembled from any fluorescence data using a spreadsheet 
program. File type determination is handled algorithmically.

Based on @menon_opsin_2011 and @goren_constitutive_2014 and largely keeping with
nomenclature used therein, data is processed as follows. Input is format checked
and defaults are injected for facultative parameters/columns as appropriate. 
Fluorescence spectra are parsed. This includes algorithmic determination of when 
dithionite was added to the sample using peak detection by continuous wavelet
transformation as implemented in the *wmtsa* R package [@percival_wavelet_2000;@constantine_wmtsa:_2013]. 
Acquisition time is aligned for all spectra in the data set such that the zero 
time point henceforth corresponds to the time of addition. 
Pre-dithionite-addition baseline fluorescence $F_{\text{baseline}}$ is 
determined for each spectrum by averaging (median) over the ten values preceding
dithionite addition. Post-dithionite-addition minimum fluorescence 
$F_{\text{min}}$ is analogously calculated from the last ten data points before 
(and including) a default 400 s or the time point of measurement supplied by the
user in the parameter spreadsheet. $F_{\text{min}}$ is volume-corrected based on
supplied reaction volumes with and without dithionite and for each spectrum/data
point a fluorescence reduction is calculated as follows:
$$
y_{\text{raw}} = 1 - \frac{F_{\text{min}}}{F_{\text{baseline}}}
$$
A relative fluorescence reduction $y$ is calculated scaled to a (required) liposomes-only/no-protein control and the probability $P_{\geq 1}$ for a 
liposome to contain at least one scramblase molecules is calculated using
$$
P_{\geq 1} = \frac{(y - y_0)}{(y_{\text{max}} - y_0)}
$$
where $y$ is the relative fluorescence reduction for the proteoliposome 
measurement and $y_0$ for a measurement of protein-free liposomes. 
$y_{\text{max}}$ refers to either the maximal $y$ in the experiment or is 
derived from a monoexponential fit to the data (default) as a precaution for the 
case where the protein/phospholipid titration did not reach the plateau of the 
saturation curve. The behavior is adjustable using the `scaleTo` parameter of 
the corresponding functions (see below). Next a monoexponential curve is fitted
to the relationship of $P_{\geq 1}$ to the protein to phospholipid ratio (PPR; 
(mg/mmol)), providing the scramblase-characterizing fit factor $\alpha$.

#Main Features
##Template Generation for Data Input
Supporting the user in providing the input data needed, the following function 
generates a spreadsheet program-compatible tab-delimited plain text template, 
including column names, commentary, expected data type and the default used when
omitted.

```{r, code=formatR::usage('scramblaseAssayInputTemplate'), eval=FALSE}
```

The application agnostic tab-delimited format must be maintained for the 
completed table, which is subjected to extensive input checks when read by 
*flippant*.

##Graphic Representation of Spectral Traces
In a first, graphical data analytic step of a scramblase assay-derived data 
set, fluorescent spectra or traces may be plotted in publication-appropriate
quality with acquisition time on the x- and fluorescence intensity on the y-axis
using

```{r, code = formatR::usage('scramblaseAssayTraces'), eval = FALSE}
```

and handing the function the path to the user assembled spreadsheet of input 
data (`x`). The `adjust` argument to the function call serves to selectively 
inactivate *wmtsa*-provided (see above) alignment of multiple spectra to the 
time of dithionite addition. Acquisition time coverage of the plot may be tuned 
by user-provided minimal (`timeMinSec` ) and maximal (`timeMaxSec`) time points
to be included (in seconds).

##Scramblase Assay Analysis
Plotting protein to phospholipid ratio (PPR) against the likelihood for a given
proteoliposome to contain one or more scramblases, the PPR plot is at the heart 
of the analysis of dithionite scramblase assays, allowing for the visual 
comparison of specific scramblase activity between interrogated experimental 
series, which may, for example, be representative of protein preparations including 
mutant protein forms or crude extracts with or without depletion of a candidate
scramblase. 

```{r, code = formatR::usage('scramblaseAssayPlot'), eval = FALSE}
```

allows for the complete generation of the PPR plot in publication quality from 
the user assembled spreadsheet of input data (`x`), including automatically 
added graphical representations the PPR value equal to the fit constant 
$\alpha$ ($P_{\geq 1} = 1 - e^{-1}$), which implies all liposomes on average 
carrying one scramblase and 63% have one or more 
[i.e. are active; @goren_constitutive_2014]. The behavior of the function may be
adjusted to use scramblase-mediated fluorescence reduction normalized to either
the maximum measurement in the series (`scaleTo = "data"`) or the plateau of an
exponential fit to it (`scaleTo = "model"`; for reasoning see above) and whether
or not to require the exponential curve fitted to the data to go through the
origin. The `splitByExperiment` argument is used to indicate whether independent
experiments are integrated into a single analysis (`splitByExperiment = FALSE`)
or treated separately (`splitByExperiment = TRUE`; see below for example usage).

Numeric scramblase assay results such as $\alpha$ may be generated in 
tabular form using

```{r, code = formatR::usage('scramblaseAssayStats'), eval = FALSE}
```

with the same modifying arguments to the function as for `scramblaseAssayPlot`.

#Case Study
*Retinitis pigmentosa* is a degenerative disease of the retina, the majority of 
cases of which is linked to mutant forms of the G-protein coupled receptor 
rhodopsin [@mendes_mechanisms_2005]. Motivated by the recent discovery that 
opsin/rhodopsin has scramblase activity [@menon_opsin_2011;@goren_constitutive_2014],
we hypothesized that a class of enigmatic rhodopsin mutations, which are known
to cause *retinitis pigmentosa* yet do not display the commonly associated 
molecular phenotypes of impaired binding of retinal, folding and/or trafficking or 
transducin activation, may in fact be impacted in their scramblase activity.
This hypothesis was disproved when tested using the dithionite scramblase assay
among other techniques. However, the same set of experiments pointed at a defect 
in rhodopsin dimerization prior to insertion into liposomes, generating the new 
hypothesis that the mutations may cause disease by interfering with the highly 
ordered quarternary structure of the molecule in the retina 
[@ploier_perturbed_submitted].

*flippant*s capabilities are demonstrated here by reanalyzing data underlying 
Figure 4C/D in @ploier_perturbed_submitted. The following work flow in a first 
step plots the corresponding spectral traces (Figure 1A). Note that algorithmic
alignment to the time of dithionite addition is employed (default) as well as 
custom trimming of the time axis to account for divergent acquisition 
times between experiments.

```{r, eval = FALSE}
# Plot spectral traces (Figure 1A)
scramblaseAssayTraces("inputTable.txt", timeMaxSec = 350)
```

```{r, echo = FALSE, warning = FALSE, message = FALSE, fig.cap = "***flippant*-based reanalysis of data in Figure 4C/D from @ploier_perturbed_submitted.** (A) represents spectral traces as produced by a call to the function `scramblaseAssayTraces`, while (B) and (C) show protein to phospholipid ration (PPR) plots for the data using a x-axis trimmed to a maximum value of 2 (see text for details). In (B) data is separately plotted for independent experiments, while (C) combines all dat points, aiming for reliability of fit."}
setwd(tmpDir)
figure1 <- addpanel(
  figure = figure1,
  topPanel = 1,
  bottomPanel = 3,
  leftPanel = 1, 
  rightPanel = 3, 
  panel = scramblaseAssayTraces("inputTable.txt", timeMaxSec = 350) + theme_bw())
figure1 <- addpanel(
  figure = figure1,
  topPanel = 4,
  bottomPanel = 5,
  leftPanel = 1, 
  rightPanel = 3, 
  panel = scramblaseAssayPlot("inputTable.txt") + 
    theme_bw() + 
    xlim(0, 2) + 
    xlab(expression("PPR"~~("mg"/"mmol"))) +
    ylab(expression(P[{phantom()>=1}])))
figure1 <- addpanel(
  figure = figure1,
  topPanel = 6,
  bottomPanel = 7,
  leftPanel = 1,
  rightPanel = 2,
  panel = scramblaseAssayPlot("inputTable.txt", splitByExperiment = FALSE) + 
    theme_bw() + 
    xlim(0, 2) + 
    xlab(expression("PPR"~~("mg"/"mmol"))) +
    ylab(expression(P[{phantom()>=1}])))
grid.draw(figure1)
```

Scramblase activity analysis as described above is performed on the data set with
the following calls.

```{r, eval = FALSE}
# Plot PPR plot (Figure 1B)
scramblaseAssayPlot("inputTable.txt") + xlim(0, 2)
```

```{r, echo = FALSE}
setwd(tmpDir)
max_PPR <- read.delim(
    "inputTable.txt",
    check.names = FALSE) %>%
  flippant:::calculatePpr() %>%
  max(
    na.rm = TRUE) %>%
  round(
    digits = 2
  )
```

The call results in the protein to phospholipid ratio (PPR) plot shown in
Figure 1B. Using the customization facilities provided by `ggplot2` 
[@wickham_ggplot2_2009] as a plotting system the plot's x-axis is trimmed to
avoid visual compression of all data by a single data point from the wild type 
rhodopsin series in experiment 1, where $\text{PPR}=`r max_PPR`$. Dashed lines
are added automatically as graphical representations of the fit constant 
$\alpha$.

Figure 1B emphasizes reproducibility between independent experiments
("Experiment 1" and "Experiment 2") on the four probed rhodopsins. Forgoing
reproducibility analysis in favor of a potentially more reliable result and
following more closely the analysis in Figure 4C/D by
@ploier_perturbed_submitted, a single fit may be performed for the combined data
from all experiments as follows, resulting in Figure 1C:

```{r, eval = FALSE}
# Plot PPR plot (Figure 1C)
scramblaseAssayPlot("inputTable.txt", splitByExperiment = FALSE) + xlim(0, 2)
```

The graphical representation of the fit constant $\alpha$ in the PPR plots is
difficult to visually evaluate and easy access to the numeric values desirable.
Recapitulation of the calculations underlying PPR plot generation for Figure 1C
by `scramblaseAssayPlot` in a call to `scramblaseAssayStats`, this is achieved
and produces tabular output represented in Table 1.

```{r, eval = FALSE}
# Write out results in tabular manner (Table 1)
scramblaseAssayStats("inputTable.txt" splitByExperiment = FALSE)
```

```{r, echo = FALSE}
setwd(tmpDir)
ppr_results <- scramblaseAssayStats("inputTable.txt", splitByExperiment = FALSE)
ppr_results_names <- ppr_results$`Experimental Series` %>%
  head(n = nrow(ppr_results) - 1) %>%
  paste(collapse = ", ") %>%
  paste("and") %>%
  paste(tail(ppr_results$`Experimental Series`, n = 1))
ppr_results_alpha_print <- ppr_results$`Fit Constant` %>%
  round(2) %>%
  head(n = nrow(ppr_results) - 1) %>%
  paste(collapse = ", ") %>%
  paste("and") %>%
  paste(tail(ppr_results$`Fit Constant`, n = 1))
knitr::kable(
  ppr_results,
  row.names = FALSE,
  caption = "**Tabular output of *flippant*-based reanalysis of data in Figure 4C/D from @ploier_perturbed_submitted.** An plain text-version of this output is produced by calling `scramblaseAssayStats`.")
```

The resulting values for the fit constant $\alpha$ correspond with
`r ppr_results_alpha_print` for `r ppr_results_names`, respectively, very well
to the prior manual analysis by @ploier_perturbed_submitted, where
$\alpha \approx 0.13$ was reported for mutant and $\alpha \approx 0.26$ for wild
type rhodopsin. Based on the manual analysis of the data we conclude in @ploier_perturbed_submitted that the mutant rhodopsins investigated are
reconstituted into proteoliposomes in monomeric form, contrary to the wild type,
which is reconstituting as a homodimer. 
*flippant* analysis of the data shows identical trends and close matches of the
results, supporting identical conclusions.

#Conclusion

```{r, echo = FALSE, cache = TRUE}
setwd(tmpDir)

analysis_timing_n <- 100

analysis_timing_function <- function(){
  scramblaseAssayTraces(
    "inputTable.txt",
    timeMaxSec = 350)
  scramblaseAssayPlot("inputTable.txt")
  scramblaseAssayPlot("inputTable.txt", splitByExperiment = FALSE)
  scramblaseAssayStats("inputTable.txt", splitByExperiment = FALSE)
}

analysis_timing <- microbenchmark(
  analysis_timing_function(),
  times = analysis_timing_n)

analysis_timing_median <- round(
  summary(
    analysis_timing)$median,
  2)

analysis_timing_iqr <- round(
  diff(
    unlist(
      summary(analysis_timing)[c("lq", "uq")],
      use.names = FALSE)),
  2)
```

With *flippant* we present an integrated solution for data analysis in the 
context of dithionite scramblase assays. Requiring only basic familiarity with 
R as an environment for statistical computing and graphics 
[@r_core_team_r:_2015], scientists can quickly analyse such data and arrive at 
publication-grade graphics that offer extensive facilities for individual 
optimization and adaptation [@wickham_ggplot2_2009]. On a laptop system with 
`r parallel::detectCores()` cores and 8 GB RAM (no SSD), running `Linux` and 
R version `r paste(R.version[c("major","minor")], collapse = ".")`, the
statements generating equivalents to the plots and calculations presented here
take a mere `r analysis_timing_median` seconds (median; 
$\text{IQR}=`r analysis_timing_iqr`~\text{s}$; $n=`r analysis_timing_n`$).

Results from *flippant*-driven analysis track prior manual analysis well. 
Deviations observed may stem from algorithmic differences such as the fitting 
routine used (*flippant* uses the Nash variant of the Marquardt 
nonlinear least squares algorithm as implemented by @nash_nlmrt:_2014) and 
general handling of significant digits and rounding.

The choice of R for the implementation is not only motivated by the extensive
tool kit for statistic methodology and exquisite plotting environment it 
provides, but also by the fact that it is a scripting language, rendering the
implementation unequivocally open to scrutiny and adaptation or extension by 
users.

*flippant*, in conclusion, provides means for quick, reproducible data analysis
in the context of dithionite scramblase analysis and a platform for review, 
dissemination and extension of the strategies it employs. As BSA-back 
extraction, a different assay for lipid translocation across liposome bilayers
[@kubelt_transbilayer_2002;@vehring_flip-flop_2007], can utilize an essentially 
identical data analysis workflow, *flippant* is expected to facilitate analysis
of such experiments as well.

#Funding
J.G., R.J.C. and the Proteomics Core at WCMC-Q are supported by ``Biomedical 
Research Program'' funds at Weill Cornell Medical College in Qatar, a program 
funded by Qatar Foundation. B.P. was supported by a grant from Qatar National 
Research Fund’s National Priority Research Program (5-669-1-112) to J.G. and
A.K.M., as well as by the Austrian Science Fund (FWF), project J3686-B21.

#Acknowledgements

#References
